### PR TITLE
Refactoring segment naming to add different schemes

### PR DESCRIPTION
**Pull request by jenniferdai**
Refactoring segment naming logic so that we can add different segment naming schemes & pass in different stats collectors.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.common.utils.SegmentNameBuilder;
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+public class SimpleSegmentNameGenerator implements SegmentNameGenerator {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: Please rename to DefaultSegmentNameGenerator, thanks

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameGenerator.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+public interface SegmentNameGenerator {
+  String getSegmentName(AbstractColumnStatisticsCollector statsCollector, SegmentNameConfig segmentNameConfig) throws Exception;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameGenerator.java]: remame var from statsCollector to timeColStatsCollector. Also, add javadocs.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  public String getSegmentName(AbstractColumnStatisticsCollector statsCollector, SegmentNameConfig config) throws Exception {
+    String timeColumn = config.getTimeColumnName();
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: final. also rename to timeColumnName

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    String segmentName;
+    if (timeColumn != null && timeColumn.length() > 0) {
+      final Object minTimeValue = statsCollector.getMinValue();
+      final Object maxTimeValue = statsCollector.getMaxValue();
+      segmentName = SegmentNameBuilder
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: Ideally we should not have SegmentNamBuilder class at all, but I think it is being used in the third-eye code, so we cannot remove it now.

File: pinot-core/src/main/java/com/linkedin/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java [hunk 5]
```diff
@@ -25,6 +25,10 @@
 import com.linkedin.pinot.core.data.readers.FileFormat;
 import com.linkedin.pinot.core.data.readers.RecordReaderConfig;
 import com.linkedin.pinot.core.indexsegment.utils.AvroUtils;
+import com.linkedin.pinot.core.segment.DefaultSegmentNameConfig;
+import com.linkedin.pinot.core.segment.DefaultSegmentNameGenerator;
+import com.linkedin.pinot.core.segment.SegmentNameConfig;
+import com.linkedin.pinot.core.segment.SimpleSegmentNameGenerator;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java]: remove

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameConfig.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.segment;
+
+/**
+ * An interface for segment name config so that we can pass in different naming schemes if we were to call this from Hadoop.
+ */
+public interface DefaultSegmentNameConfig {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameConfig.java]: rename this as the interface SegmentNameConfig

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+/**
+ * An interface that allows generates names for segments depending on the naming scheme.
+ */
+public interface DefaultSegmentNameGenerator {
+  String getSegmentName(AbstractColumnStatisticsCollector timeColStatsCollector, DefaultSegmentNameConfig segmentNameConfig) throws Exception;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java]: change signature to 
public String getSegmentName(AbstractColumnStatisticsColelctor timeColStatsCollector, SegmentNameConfig segmentNameConfig); 

Do we really want to throw "Exception"? If there are specific exceptions, then list them there. Do not throw high level Exception as far as possible

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.segment;
+
+/**
+ * DO NOT remove any getters, setters, or fields from this class otherwise it will be backwards incompatible.
+ */
+public class SegmentNameConfig implements DefaultSegmentNameConfig {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: change to "public class DefaultSegmentNameConfig implements SegmentNameConfig"

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  private String _timeColumnName;
+  private String _tableName;
+  private String _segmentNamePostfix;
+  private String _segmentName;
+
+  public SegmentNameConfig() {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: do we need this ctor? Minimize number of ctors unless absolutely needed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+  }
+  public SegmentNameConfig(String timeColumnName, String tableName, String segmentNamePostfix, String segmentName) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: do not include segmentName in the constructor. Make it more difficult to use this by having a deprecated setter.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    _tableName = tableName;
+    _segmentNamePostfix = segmentNamePostfix;
+    _segmentName = segmentName;
+  }
+
+  /**
+   * We have this here because this option is available in the current SegmentGeneratorConfig, and we don't want to take
+   * away this functionality in case users are setting their own segment names.
+   */
+  public void setSegmentName(String segmentName) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: annotate with @deprecated, @discouraged (if there is anything like that). Definitely document that this optioin will be removed soon, and must NOT be used.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+  public String getSegmentName() {
+    return _segmentName;
+  }
+
+  public String getTimeColumnName() {
+    return _timeColumnName;
+  }
+
+  public void setTimeColumnName(String timeColumnName) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: not needed

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    _timeColumnName = timeColumnName;
+  }
+
+  public String getTableName() {
+    return _tableName;
+  }
+
+  public void setTableName(String tableName) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: not needed

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    _tableName = tableName;
+  }
+
+  public String getSegmentNamePostfix() {
+    return _segmentNamePostfix;
+  }
+
+  public void setSegmentNamePostfix(String segmentNamePostfix) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SegmentNameConfig.java]: Needed? Not sure.

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+/**
+ * An interface that allows generates names for segments depending on the naming scheme.
+ */
+public interface DefaultSegmentNameGenerator {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java]: change interface name to SegmentNameGenerator

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.common.utils.SegmentNameBuilder;
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+public class SimpleSegmentNameGenerator implements DefaultSegmentNameGenerator {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: change class name to DefaultSegmentNameGenerator

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  public String getSegmentName(AbstractColumnStatisticsCollector statsCollector, DefaultSegmentNameConfig config) throws Exception {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: do not add throws exception in signature

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentIndexCreationDriverImpl.java [hunk 18]
```diff
@@ -414,21 +414,9 @@ public void buildRaw() throws Exception {
...
-      if (timeColumn != null && timeColumn.length() > 0) {
-        final Object minTimeValue = statsCollector.getColumnProfileFor(timeColumn).getMinValue();
-        final Object maxTimeValue = statsCollector.getColumnProfileFor(timeColumn).getMaxValue();
-        segmentName = SegmentNameBuilder
-            .buildBasic(config.getTableName(), minTimeValue, maxTimeValue, config.getSegmentNamePostfix());
-      } else {
-        segmentName = SegmentNameBuilder.buildBasic(config.getTableName(), config.getSegmentNamePostfix());
-      }
-    }
+    String timeColumn = config.getTimeColumnName();
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/SegmentIndexCreationDriverImpl.java]: final

File: pinot-core/src/test/java/com/linkedin/pinot/core/segment/SegmentNameGeneratorTest.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+@Test
+public class SegmentNameGeneratorTest {
+
+  public void testPostfix() throws Exception {
+    ColumnMetadataTest columnMetadataTest = new ColumnMetadataTest();
+    // Build the Segment metadata.
+    SegmentGeneratorConfig config = columnMetadataTest.CreateSegmentConfigWithoutCreator();
+    SegmentNameConfig segmentNameConfig = new SegmentNameConfig("daysSinceEpoch", "mytable", "postfix", null);
+    config.setSegmentNameConfig(segmentNameConfig);
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/com/linkedin/pinot/core/segment/SegmentNameGeneratorTest.java]: also add a test that does not call setSegmentNameConfig() we should be able to use the DefaultSegmentNameConfig in these cases

File: pinot-core/src/main/java/com/linkedin/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java (hunk 5 above)
Comment (jenniferdai) [inline pinot-core/src/main/java/com/linkedin/pinot/core/indexsegment/generator/SegmentGeneratorConfig.java]: I believe I'm using these!

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java (hunk 7 above)
Comment (jenniferdai) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java]: statsCollector.getMinValue() throws the high level exception

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java (hunk 17 above)
Comment (jenniferdai) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/SimpleSegmentNameGenerator.java]: statsCollector.getMinValue() throws a high level exception

File: pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+ */
+package com.linkedin.pinot.core.segment;
+
+import com.linkedin.pinot.common.data.Schema;
+import com.linkedin.pinot.common.utils.SegmentNameBuilder;
+import com.linkedin.pinot.core.segment.creator.AbstractColumnStatisticsCollector;
+
+
+public class DefaultSegmentNameGenerator implements SegmentNameGenerator {
+  String _segmentName;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGenerator.java]: private final for all vars please

File: pinot-core/src/test/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGeneratorTest.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+  public void testPostfix() throws Exception {
+    ColumnMetadataTest columnMetadataTest = new ColumnMetadataTest();
+    // Build the Segment metadata.
+    SegmentGeneratorConfig config = columnMetadataTest.CreateSegmentConfigWithoutCreator();
+    SegmentNameGenerator segmentNameGenerator = new DefaultSegmentNameGenerator("daysSinceEpoch", null, "mytable", "1");
+    config.setSegmentNameGenerator(segmentNameGenerator);
+    SegmentIndexCreationDriver driver = SegmentCreationDriverFactory.get(null);
+    driver.init(config);
+    driver.build();
+     Assert.assertEquals(driver.getSegmentName(), "mytable_1756015683_1756015683_1");
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/com/linkedin/pinot/core/segment/DefaultSegmentNameGeneratorTest.java]: nit: indent

File: pinot-core/src/test/java/com/linkedin/pinot/core/segment/index/ColumnMetadataTest.java [hunk 22]
```diff
@@ -67,6 +67,17 @@ public SegmentGeneratorConfig CreateSegmentConfigWithoutCreator()
     return config;
   }
 
+  public SegmentGeneratorConfig createSimpleSegmentConfig()
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/com/linkedin/pinot/core/segment/index/ColumnMetadataTest.java]: Can you move this method to DefaultSegmentNameGeneratorTest? the avro file should be accessable from there as well.

**Pull request merged**